Multi search (#225)
meilisearch/specifications: text/0192-multi-search-api.md

# Multi-search API

## 1. Summary

> The multi-search endpoint performs a batch of search queries as if using the [search API](./0118-search-api.md) for an index.
> 
> This endpoint allows, for example:
> 
> - Execute different search queries on the same index
> - Execute search queries on different indexes
> 

Reviewer: ```suggestion
The multi-search endpoint performs multiple search queries on one or more indexes by bundling them into a single request.
```

Each search query has its own results set.

## 2. Motivation

- Perform multiple queries in a single HTTP request

## 3. Functional Specification

Meilisearch exposes 1 route to perform multi-search requests:

- POST `/multi-search`

If a master key is used to secure a Meilisearch instance, the auth layer returns the following errors:

- 🔴 Accessing these routes without the `Authorization` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- 🔴 Accessing these routes with a key that does not have permissions (i.e. other than the master key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.

> If any of the search queries fail to execute, the response returns the corresponding error instead of the array of results. If multiple queries would have failed, only the first encountered failure is returned.

Reviewer: ```suggestion
If any of the search queries fail to execute, the response returns the corresponding error instead of the array of results. If multiple queries fail, only the first encountered failure is returned.
```

# Multi-search API

## 1. Summary

> The multi-search endpoint performs a batch of search queries as if using the [search API](./0118-search-api.md).

Reviewer 2: ```suggestion
The multi-search endpoint performs a batch of search queries as if using the [search API](./0118-search-api.md) for an index.

This endpoint allows, for example:

- Execute different search queries on the same index
- Execute the same search query on different indexes

Each search query has its own results set.
```

- POST `/multi-search`

If a master key is used to secure a Meilisearch instance, the auth layer returns the following errors:

- 🔴 Accessing these routes without the `Authorization` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- 🔴 Accessing these routes with a key that does not have permissions (i.e. other than the master key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.

If any of the search queries fail to execute, the response returns the corresponding error instead of the array of results.

`POST` HTTP verb errors:

- 🔴 Omitting the Content-Type header returns a [missing_content_type](0061-error-format-and-definitions.md#missing_content_type) error.
- 🔴 Sending an empty Content-Type returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- 🔴 Sending a different Content-Type than `application/json` returns an [invalid_content_type](0061-error-format-and-definitions.md#invalid_content_type) error.
- 🔴 Sending an empty payload returns a [missing_payload](0061-error-format-and-definitions.md#missing_payload) error.

> - 🔴 Sending an invalid JSON payload returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.

Reviewer 2: ```suggestion
- 🔴 Sending an invalid JSON payload returns a [malformed_payload](0061-error-format-and-definitions.md#malformed_payload) error.
- 🔴 Sending an inexistent `indexUid` in a query object returns an [index_not_found](0061-error-format-and-definitions.md#index_not_found) error.
- 🔴 Sending an invalid format for the `indexUid` property in a query object returns an [invalid_index_uid](0061-error-format-and-definitions.md#invalid_index_uid) error.
```

# Multi-search API

## 1. Summary

The multi-search endpoint performs a batch of search queries as if using the [search API](./0118-search-api.md) for an index.

This endpoint allows, for example:

- Execute different search queries on the same index

> - Execute the same search query on different indexes

Reviewer 3: ```suggestion
- Execute search queries on different indexes
```

Should we just say that you can perform multiple searches on different indexes decorrelated of the search query in itself?

Since these are example, my comment might not be relevant. My first intuition though was that it was either one or the other and thus I had to use the same search query when querying on multiple indexes.

@maryamsulemani97 what do you think?

Author: I'm OK with the suggestion.

Each search query has its own results set.

## 2. Motivation

- Perform multiple queries in a single HTTP request

## 3. Functional Specification

Meilisearch exposes 1 route to perform multi-search requests:

- POST `/multi-search`

If a master key is used to secure a Meilisearch instance, the auth layer returns the following errors:

- 🔴 Accessing these routes without the `Authorization` header returns a [missing_authorization_header](0061-error-format-and-definitions.md#missing_authorization_header) error.
- 🔴 Accessing these routes with a key that does not have permissions (i.e. other than the master key) returns an [invalid_api_key](0061-error-format-and-definitions.md#invalid_api_key) error.

> If any of the search queries fail to execute, the response returns the corresponding error instead of the array of results.

Reviewer 3: ```suggestion
If any of the search queries fail to execute, the response returns the corresponding error instead of the array of results. If the queries failed for multiple reasons, only the first discovered issue is returned.
```

Author: I think I'd rather go:

```suggestion
If any of the search queries fail to execute, the response returns the corresponding error instead of the array of results. If multiple queries would have failed, only the first encountered failure is returned.
```

Reviewer 3: way better ✨

- 🔴 Sending an invalid format for the `indexUid` property in a query object returns an [invalid_index_uid](0061-error-format-and-definitions.md#invalid_index_uid) error.

### 3.1. Search Payload Parameters

| Field        | Type             | Required |
|--------------|------------------|----------|
| [`queries`](#311-queries)  | Array of Objects | True     |

#### 3.1.1. `queries`

- Type: Object
- Required: True

The `queries` object contains the list of search queries to perform.

Each element of this array is a JSON object with the required field `indexUid`, the uid of the index to be searched. Other fields of this object are optional, and identical to the ones in the [existing search routes](./0118-search-api.md#31-search-payload-parameters) (`q`, `limit`, etc.).

### 3.2. Search Response Properties

| Field                     | Type          | Required |
|---------------------------|---------------|----------|
| [`results`](#321-results) | Array[Object] | True     |

#### 3.2.1. `results`

- Type: Array[Objets]
- Required: True

Results of the search queries as an array of search results.

> Each element of this array is a JSON object representing the results of the search query with the same index in the `queries` array from the request, and an additional `indexUid` field that is equal to the `indexUid` field in that search query.

Reviewer 3: This sentence is a bit confusing. I would say something in the lines of:

```
Each element in this array contains the results of the search queries in the same order they have been requested. Additionally to the usual fields returned by a search result, see [`/search`](#link), an `indexUid` field is present with the index UID on which the search has been performed.

example:

Search requests:
\`\`\`json
{
   "queries": [ { "indexUid": "movie", "q": "wonder" }, { "indexUid": "books", "q": "king" } ]
}
\'\'\'

Search response:
\`\`\`json
{
   "results": [ 
      {
         "indexUid": "movie",
         "hits": [ { "title": "wonderwoman" } ],
         // other search results fields: processingTimeMs, limit, ...
      },
      {
         "indexUid": "books",
         "hits": [ { "title": "king kong theory" } ],
         // other search results fields: processingTimeMs, limit, ...
      },
   ]
}
\`\`\`
```

If you agree, you can change the example if you want 😂

Author: sure

btw, if you don't want to escape the ```, you can use code fences of a different length to nest them:

`````
````
```
I used a code fence of size 5: ````` so that I can nest code fences of sizes 3 and 4 inside of them!
```
````
`````

Author: So, as a suggestion:

````suggestion
Each element in this array contains the results of the search queries in the same order they have been requested. Additionally to the [usual fields returned by a search result](./0118-search-api.md#31-formatting-search-results), an `indexUid` field is present with the index UID on which the search has been performed.

example:

Search requests:
```json
{
   "queries": [ { "indexUid": "movie", "q": "wonder" }, { "indexUid": "books", "q": "king" } ]
}
'''

Search response:
```json
{
   "results": [ 
      {
         "indexUid": "movie",
         "hits": [ { "title": "wonderwoman" } ],
         // other search results fields: processingTimeMs, limit, ...
      },
      {
         "indexUid": "books",
         "hits": [ { "title": "king kong theory" } ],
         // other search results fields: processingTimeMs, limit, ...
      },
   ]
}
```
````

Reviewer 3: 👌 perfect

Results of the search queries as an array of search results.

Each element in this array contains the results of the search queries in the same order they have been requested. Additionally to the [usual fields returned by a search result](./0118-search-api.md#31-formatting-search-results), an `indexUid` field is present with the index UID on which the search has been performed.

example:

Search queries:
```json
{
   "queries": [ { "indexUid": "movie", "q": "wonder" }, { "indexUid": "books", "q": "king" } ]
}

> '''

Reviewer 3: These are the wrong quotes. I don't manage to make a suggestion with the correct ones

Author: ````suggestion
```
````

wow didn't notice they were escaped ticks rather than backticks in your initial comment 😅 

this should fix it
